routing/DELEGATED_ROUTING_HTTP.md

```diff
@@ -0,0 +1,101 @@
+# ![](https://img.shields.io/badge/status-wip-orange.svg?style=flat-square) Delegated Routing HTTP API
+
+**Author(s)**:
+- Gus Eggert
+
+**Maintainer(s)**:
+
+* * *
+
+**Abstract**
+
+"Delegated routing" is a mechanism for IPFS implementations to use for offloading content routing to another process/server. This spec describes an HTTP API for delegated routing.
+
+# Organization of this document
+
+- [Introduction](#introduction)
+- [Spec](#spec)
+  - [Interaction Pattern](#interaction-pattern)
+  - [Cachability](#cachability)
+  - [Transports](#transports)
+  - [Protocol Message Overview](#protocol-message-overview)
+    - [Known Methods](#known-methods)
+- [Method Upgrade Paths](#method-upgrade-paths)
+- [Implementations](#implementations)
+
+# API Specification
+By default, the Delegated Routing HTTP API uses the `application/json` content type. Clients and servers may optionally negotiate other content types such as `application/cbor`, `application/vnd.ipfs.rpc+dag-json`, etc. using the standard `Accept` and `Content-Type` headers.
+
+- `GET /v1/providers/{CID}`
+    - Reframe equivalent: FindProviders
+    - Response
+        
+        ```json
+        {
+        	"Providers": [
+        		{
+        			"PeerID": "...",
+        			"Multiaddrs": ["...", "..."]
+        			"Protocols": [
+        				{
+        					"Codec": 2320,
+        					"Payload": <opaque data>
+        				}
+        			]
+        		}
+        	]
+        	"NextPageToken": "<token>"
+        }
+        ```
+        
+    - Default limit: 100 providers
+    - Optional query parameters
+        - `transfer` only return providers who support the passed transfer protocols, expressed as a comma-separated list of multicodec IDs such as `2304,2320`,
+        - `transport` only return providers whose published multiaddrs explicitly support the passed transport protocols, such as `/quic` or `/tls/ws`.
```

Comment:
I'd need to look into this carefully, I don't think the multiaddr spec requires knowledge of multicodecs, we could e.g. just search for a multiaddr component that matches the passed one, instead of mapping components to multicodec codes which means this would need to know the multicodec code for each component, which sounds brittle.

Reply:
> multiaddr spec requires knowledge of multicodecs

every [protocol we support in multiaddr](https://github.com/multiformats/multiaddr#protocols) is listed in [protocols.csv](https://github.com/multiformats/multiaddr/blob/master/protocols.csv) and every code added there  MUST also have a code reserved in the [table.csv](https://github.com/multiformats/multicodec/blob/master/table.csv) – this is the main purpose of